pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java

```diff
@@ -411,10 +411,28 @@ public UnAckedMessageTracker getUnAckedMessageTracker() {
         return unsubscribeFuture;
     }
 
+    @Override
+    public void initReceiverQueueSize() {
+        if (conf.isAutoScaledReceiverQueueSizeEnabled()) {
+            // turn on autoScaledReceiverQueueSize
+            int size = Math.min(INITIAL_RECEIVER_QUEUE_SIZE, maxReceiverQueueSize);
+            if (batchReceivePolicy.getMaxNumMessages() > 0) {
+                // consumerImpl may store (half-1) permits locally.
+                size = Math.max(size, 2 * batchReceivePolicy.getMaxNumMessages() - 2);
```

Comment:
If `batchReceivePolicy.getMaxNumMessages()` is close or equal to `maxReceiverQueueSize`, will `size` exceed the  limit of `ConsumerBuilderImpl.receiverQueueSize()`?

Reply:
> If `batchReceivePolicy.getMaxNumMessages()` is close or equal to `maxReceiverQueueSize`, will `size` exceed the limit of `ConsumerBuilderImpl.receiverQueueSize()`?

@aloyszhang Yes, it should exceed the limit if we expect it to consume normally. And this should be a pre-existing issue that consumer would get stuck if `batchReceivePolicy.getMaxNumMessages()` is equal to `receiverQueueSize`